control/src/robocup/radio/RadioLink.cpp

```diff
@@ -14,93 +19,141 @@ void RadioLink::init() {
     radioInitialized = radio->isInitialized();
 }
 
-void RadioLink::send(const BatteryVoltage& batteryVoltage,
-                     const FPGAStatus& fpgaStatus,
-                     const KickerInfo& kickerInfo,
-                     const RobotID& robotID,
-                     DebugInfo& debugInfo) {
-    std::array<uint8_t, rtp::ReverseSize> packet;
-    rtp::Header* header = reinterpret_cast<rtp::Header*>(&packet[0]);
-    rtp::RobotStatusMessage* status = reinterpret_cast<rtp::RobotStatusMessage*>(&packet[rtp::HeaderSize]);
-
-    header->address = rtp::BASE_STATION_ADDRESS;
-    header->port    = rtp::PortType::CONTROL;
-    header->type    = rtp::MessageType::CONTROL;
-
-    uint8_t motorErrors = 0;
-    for (int i = 0; i < 4; i++)
-        motorErrors |= static_cast<int>(fpgaStatus.motorHasErrors[i]) << i;
-
-    status->uid             = robotID.robotID;
-    status->battVoltage     = batteryVoltage.rawVoltage;
-    status->motorErrors     = motorErrors;
-    status->ballSenseStatus = static_cast<uint8_t>(kickerInfo.ballSenseTriggered);
-    status->kickStatus      = static_cast<uint8_t>(kickerInfo.kickerCharged);
-    status->kickHealthy     = static_cast<uint8_t>(kickerInfo.kickerHasError);
-    status->fpgaStatus      = static_cast<uint8_t>(fpgaStatus.FPGAHasError);
-
-#if 0
-    status->num_frames = 0;
-
-    for (int i = 0; i < debugInfo.num_debug_frames && status->num_frames < rtp::MAX_DEBUG_FRAMES; i++) {
-        auto& to = status->debug_frames[status->num_frames];
-        const auto& from = debugInfo.debug_frames[i];
-
-        to.time_ticks = from.ticks;
-        to.gyro_z = from.gyro_z;
-        to.accel_x = from.accel_x;
-        to.accel_y = from.accel_y;
-        std::copy(std::begin(from.filtered_velocity),
-                  std::end(from.filtered_velocity),
-                  std::begin(to.filtered_velocity));
-        std::copy(std::begin(from.motor_outputs),
-                  std::end(from.motor_outputs),
-                  std::begin(to.motor_outputs));
-        std::copy(std::begin(from.encDeltas),
-                  std::end(from.encDeltas),
-                  std::begin(to.encDeltas));
-
-        status->num_frames++;
+// Keep these variables in static memory. We'd rather get link errors if they're too big.
+// We only ever use one of these at a time
+static union {
+    rj_radio_RobotStatus robot_status_proto;
+    rj_radio_RobotCommand robot_command_proto;
+};
+// Block off a huge chunk of memory. It's really only this big because of (optional) log messages.
+static std::array<uint8_t, 16 * 1024> buffer{};
```

Comment:
Buffer memory, for visibility. The RobotStatus struct isn't small either (~5KiB), just because of the potential existence of log messages. Note that that does not mean that the wire size is 5KiB, but the in-memory version statically allocates the maximum size of 256B of data per log message * 16 log messages (intentional feature, avoids dynamic malloc).